### PR TITLE
Fix hard-coded byte alignment in linker scripts

### LINKED ISSUE #69859: Hard-coded 4-byte alignment in linker script overrides 8-byte alignment for 64-bit platforms
**Description**

Now there are hard-coded **4** `SUBALIGN` attribute for sections in almost all linker scripts. F.e: there is the following code in the `include/zephyr/linker/common-ram.ld`:

```
...
	ITERABLE_SECTION_RAM_GC_ALLOWED(k_queue, 4)
	ITERABLE_SECTION_RAM_GC_ALLOWED(k_fifo, 4)
	ITERABLE_SECTION_RAM_GC_ALLOWED(k_lifo, 4)
...
``` 
 during the build it provides something like that in generated `linker.cmd` 
```
...
 k_queue_area : SUBALIGN(4) { _k_queue_list_start = .; *(SORT_BY_NAME(._k_queue.static.*)); _k_queue_list_end = .; } > RAM
 k_fifo_area : SUBALIGN(4) { _k_fifo_list_start = .; *(SORT_BY_NAME(._k_fifo.static.*)); _k_fifo_list_end = .; } > RAM
 k_lifo_area : SUBALIGN(4) { _k_lifo_list_start = .; *(SORT_BY_NAME(._k_lifo.static.*)); _k_lifo_list_end = .; } > RAM
...
```  

There is the recent change in behaviour of ld linker with `SUBALIGN` attribute (see the patch https://github.com/bminor/binutils-gdb/commit/60856b5cdafc520538688258a8ce4ae8be2a8477), that may introduce some issues with building of zephyr-based applications for 64-bit platforms.

For example, let's try to build the following application for some RISC-V 64 bit platform  with `ld` of **prior vesrion**
```
#include <zephyr/kernel.h>
#include <zephyr/console/console.h>
#include <string.h>
#include <stdio.h>

int main(void)
{
	console_getline_init();

	printk("Success to invoke console_getline_init");

	return 0;
}
```

```
$ export CROSS_COMPILE=${PATH_TO_RISCV_TOOLCHAIN}/riscv-gcc/bin/riscv64-unknown-elf-
$ export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
$ west build -p -b rv64-platform samples/arch/test
$ ${PATH_TO_RISCV_TOOLCHAIN}/riscv-gcc/bin/riscv64-unknown-elf-objdump -h build/zephyr/zephyr.elf | grep k_fifo
 13 k_fifo_area   00000060  000000000000cce8  000000000000cce8  00005fa0  2**3
```

We can see that the symbol `k_fifo_area` aligned by 8 despite an attribute `SUBALIGN(4)`. 

With **new** version of `ld` it will be different 

```
$ ${PATH_TO_RISCV_TOOLCHAIN}/riscv-gcc/bin/riscv64-unknown-elf-objdump -h build/zephyr/zephyr.elf | grep k_fifo
 13 k_fifo_area   00000060  000000000000cce4  000000000000cce4  00005f9c  2**2
```

We can see that the symbol `k_fifo_area` aligned by 4. Moreover, its fixed address is `0xcce4`

The compiler doesn't know about these alignment tricks, so with this symbol it's able to generate an instruction with 8-bytes alignment requirement. In this example it generates the instruction `amoadd.d.aq a0,a4,(a5)` that causes trap. So, test fails.

### PR DESCRIPTION
**Pull request by peter-mitsis**
Most instances of the ITERABLE_SECTION_xxx() macros used a hard-coded value of 4 for the sub-alignment. As that has the potential to cause issues with 64-bit platforms that should use 8 instead, those cases now use the new macro Z_LINK_ITERABLE_SUBALIGN instead as it is defined as 4 for 32-bit platforms and 8 for 64-bit platforms.

Fixes #69859

### CODE REVIEW

File: include/zephyr/linker/iterable_sections.h [hunk 1]
```diff
@@ -32,6 +32,12 @@
 	*(SORT_BY_NAME(._##struct_type.static.*)); \
 	_CONCAT(_##struct_type, _list_end) = .
 
+#ifdef CONFIG_64BIT
+#define Z_LINK_ITERABLE_SUBALIGN  8
+#else
+#define Z_LINK_ITERABLE_SUBALIGN  4
+#endif
```
Comment (pdgendt) [inline include/zephyr/linker/iterable_sections.h:39]: Alternatively a `kconfig` symbol could be created directly

```
config ITERABLE_SECTION_SUBALIGN
	int
	default 8 if 64BIT
	default 4
```

**Comment by glarsennordic** [comment]
`conn_mgr` changes LGTM as long as tests pass. I am neutral between the KConfig and the C Macro approach.

Part of me wonders if we should look into removing the argument, or providing versions of the API that don't need the argument, but that _**would**_ be a breaking API change, so probably better not to do that for now. (Not to mention the singular iterable section in Zephyr (that I cold find) that uses an align other than 4, `zbus_channel_observation_mask`)

```c
ITERABLE_SECTION_RAM(zbus_channel_observation_mask, 1)
```

**Review by glarsennordic — DISMISSED** [review]

**Comment by glarsennordic** [comment]
Still LGTM, not approving so that others get a chance to review

**Comment by pdgendt** [comment]
> Still LGTM, not approving so that others get a chance to review

I think the CI failures should be looked at and get fixed too.

**Review by SeppoTakalo — DISMISSED** [review]

**Review by tmon-nordic — DISMISSED** [review]

**Comment by pdgendt** [comment]
> Adding the DNM tag to ensure that no one merges this (yet). There is something weird going on with many of the logging tests on 64-bit platforms with this change set. I am trying to track down the cause.

Not sure if this is for the logging tests only, `tests/misc/iterable_sections` also fails on `native_sim/native/64` but not on `native_sim`.
My guess would be that there is some mismatch in where padding is added in the linker vs compiler. I had similar issues when trying to put `__packed` structs in linker sections.

**Comment by pdgendt** [comment]
Upon closer inspection, the issue is actually in the `TYPE_SECTION_FOREACH` iterator, as it doesn't know about the struct alignment.

**Comment by pdgendt** [comment]
Maybe we can add the alignment at the end of the struct area:

```c
#ifdef CONFIG_64BIT
#define Z_LINK_ITERABLE_SUBALIGN  8
#define Z_WORD(x) QUAD(x)
#else
#define Z_LINK_ITERABLE_SUBALIGN  4
#define Z_WORD(x) LONG(x)
#endif
// ...

#define ITERABLE_SECTION_ROM(struct_type, subalign) \
	SECTION_PROLOGUE(struct_type##_area,,SUBALIGN(subalign)) \
	{ \
		Z_LINK_ITERABLE(struct_type); \
		Z_WORD(subalign) \
	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
```

Which then can be accessed using the `_list_end` address, and compute the iterator increments based on the size and alignment?

**Comment by peter-mitsis** [comment]
There was a discrepancy between linker and compiler alignment for two of the structures in the previously observed failing CI tests. On 64-bit platforms, the linker was using 8-byte alignment, but the compiler was using 4-byte alignment. Consequently, when the TYPE_SECTION_FOREACH() was used, its iterator pointer was not pointing at the correct address when it was advanced to the next element.

This updated patch set should fix that by forcing the size of those two structures to be mulitples of 8 bytes, thereby disallowing any padding between successive instances of the structures within the relevant iterable sections.

**Review by SeppoTakalo — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: include/zephyr/linker/iterable_sections.h (hunk 1 above)
Comment (pdgendt) [reply include/zephyr/linker/iterable_sections.h:39]: See #72373

**Pull request merged**